adr/ADR-32.md

```diff
@@ -0,0 +1,117 @@
+| Status   | Approved                                                    |
+| Tags     | jetstream, client                                              |
+
+## Context and Problem Statement
+
+Consuming messages from a JetStream require a large number of options and design
+decisions from client API users:
+
+- Current JetStream clients create and update a consumer definition on the fly
+  as `subscribe()` or some other functionality for consuming messages is
+  invoked. This leads to some unexpected behaviors as different clients,
+  possibly written using different versions using different options of the
+  library attempt to consume from the same consumer.
+
+- Clients implementing JetStream code are also confronted with a choice on
+  whether they should be implementing a `Pull` or `Push` subscriber.
+
+The goal of this ADR is to provide a simpler API to JetStream users that reduces
+the number of options they are confronted with and provides the expected
+performance.
+
+## Design
+
+JetStream simplification revolves around the concept of the `Consumer`. The
+consumer definition managed by the server is created by tools (such as nats cli)
+or code. A `Consumer` can be created or added as per the client language
+implementation, typically by providing a stream name and some consumer options.
+
+Clients wanting to consume messages from the consumer, simply locate the
+`Consumer`, which then provides functionality for processing, or consuming
+messages. The representation of the `Consumer` is retrieved, and provides
+operations to `fetch` and `consume` messages.
+
+### Consumers
+
+Consumer are JetStream API entities from which messages can be read. Consumers
+can be created using language specific verbs. In some cases some of the
+libraries will hang them under a _consumer api_ to prevent a collision with
+functionality that may already exist in the client for purposes of managing the
+low level `ConsumerInfo`.
+
+This means that in some languages `add` may not be available (unless major
+version bump in the library). For that reason libraries can choose how to
+instantiate a consumer:
+
+- `new Consumer(consumerInfo)`
+- `add(streamName, consumerOptions)`
+- `create(streamName, consumerOptions)`
+- `get(streamName, consumerName)`
+
+Some of the libraries can provide this functionality by chaining if that is
+appropriate, and makes sense given their current JetStreamManager implementation
+semantics:
+
+`getStream(name).getConsumer(name)`
+
+### Ordered Consumer
+
+By their very nature, OrderedConsumers don’t play well with the above API for
+retrieving the consumer, because the consumer itself will be created and
+possibly re-created by the client during it's lifecycle. Clients may wish to
+provide a:
+
+`getOrderedConsumer(streamName, startOptions, bufferingOptions)` which allows
+consuming messages while conforming to the API operations on a consumer outlined
+below.
+
```

Comment:
Does this mean we need to prevent ordered consumers in the new/add/create?
Also, maybe it's better we present Consumer as _ConsumerContext_ 
1) In Java, the name Consumer is already taken, so I'm using ConsumerContext anyway
2) It decouples the actual consumer/info from Consumer/ConsumerContext so ordered is just a variety of Consumer/Context. Also leaves the door open for other things we have not thought of yet.

Reply:
you can create them, but the object that you return must be able to be recreated at a different sequence.